src/main/scala/com/fulcrumgenomics/bam/SplitBamByReadGroup.scala

```diff
@@ -0,0 +1,113 @@
+    |splitting by the library.  All reads without a read group, or without a library when splitting by library, will be
+    |written to `<output-prefix>.unknown.bam`.
+  """)
+class SplitBamByReadGroup
+( @arg(flag='i', doc="Input SAM or BAM file.") val input: PathToBam,
+  @arg(flag='o', doc="Output prefix for all SAM or BAM files (ex. output/sample-name).") val output: PathPrefix,
+  @arg(flag='l', doc="Split by library instead of read group") val byLibrary: Boolean = false,
+  @arg(flag='u', doc="The name to use for the unknown file") val unknown: String = "unknown"
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val in       = SamReaderFactory.make().open(input)
+    val progress = ProgressLogger(logger)
+    val unknownBamAndWriter = {
+      val unknownBam: PathToBam = toOutput(unknown)
+      val unknownWriter: SAMFileWriter = new SAMFileWriterFactory().makeWriter(in.getFileHeader, true, toOutput(unknown).toFile, null)
+      WriterInfo(name=unknown, bam=unknownBam, writer=unknownWriter)
+    }
+    val writers  = createWriters(header=in.getFileHeader, unknownBamAndWriter=unknownBamAndWriter)
+    val counter  = new SimpleCounter[WriterInfo]()
+
+    in.iterator().foreach { rec =>
+      val info = writers(rec.getReadGroup)
+      info.writer.addAlignment(rec)
+      counter.count(info)
+      progress.record(rec)
+    }
+
+    writers.values.foreach(_.close())
+    unknownBamAndWriter.close()
+    in.safelyClose()
+
+    counter.toSeq sortBy { _._1.name } foreach { case (info, count) =>
+      logger.info(s"Wrote $count records to ${info.bam.getFileName}")
+    }
+  }
+
+  /** Stores the path to the BAM and the associated writer */
+  private case class WriterInfo(name: String, bam: PathToBam, writer: SAMFileWriter) {
+    def close(): Unit = writer.close()
+  }
+
+  /** Initializes the writers and returns a map from each read group to a writer. */
+  private def createWriters(header: SAMFileHeader, unknownBamAndWriter: WriterInfo): Map[SAMReadGroupRecord, WriterInfo] = {
+    val toWriter = if (byLibrary) {
+      header.getReadGroups.toSeq.groupBy { rg => rg.getLibrary }
+        .flatMap { case (library, readGroups) =>
+          val bam = toOutput(library)
+          val writer = new SAMFileWriterFactory().makeWriter(header, true, bam.toFile, null)
+          readGroups.map { rg => rg -> WriterInfo(name=library, bam=bam, writer=writer) }
+        }
+    }
+    else {
+      header.getReadGroups.map { rg =>
+        val bam = toOutput(rg.getId)
+        val writer = new SAMFileWriterFactory().makeWriter(header, true, bam.toFile, null)
+        rg -> WriterInfo(name=rg.getId, bam=bam, writer=writer)
+      }
+    }
+
+    toWriter.toMap.withDefaultValue(unknownBamAndWriter)
+  }
+
+  private[bam] def toOutput(name: String): PathToBam = {
+    val outputDir = output.getParent
+    val prefix    = PathUtil.basename(output)
```

Comment:
I think you want just `output.getFileName` here.  `PathUtil.basename` will also remove any extension.  I.e. if I say I want my prefix to be `/tmp/x.by_lib` it will strip off the `by_lib` which I wouldn't expect.